File: baekjoon_all/15000+/boj_15683.java
// Solve 2024-02-19
// Update 2025-10-14

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class boj_15683 {

    static class CCTV {
        int x, y, type;

        CCTV(int x, int y, int type) {
            this.x = x;
            this.y = y;
            this.type = type;
        }
    }

    static int n, m;
    static int[][] board;
    static int cctvCount = 0;
    static CCTV[] cctvs = new CCTV[8];
    static int[] dx = { -1, 0, 1, 0 };
    static int[] dy = { 0, 1, 0, -1 };
    static int[][][] combs = {
        {},
        {{0}, {1}, {2}, {3}},
        {{0, 2}, {1, 3}},
        {{0, 1}, {1, 2}, {2, 3}, {3, 0}},
        {{0, 1, 2}, {1, 2, 3}, {2, 3, 0}, {3, 0, 1}},
        {{0, 1, 2, 3}}
    };
    static int minHiddenArea;

    static int change(int x, int y, int d, int val, int newVal) {
        int ret = 0;
        x += dx[d];
        y += dy[d];

        while (x >= 0 && x < n && y >= 0 && y < m && board[x][y] != 6) {
            if (board[x][y] == val) {
                board[x][y] = newVal;
                ret++;
            }

            x += dx[d];
            y += dy[d];
        }

        return ret;
    }

    static void recur(int depth, int hiddenArea) {
        if (depth == cctvCount) {
            minHiddenArea = Math.min(minHiddenArea, hiddenArea);
            return;
        }

        CCTV curCctv = cctvs[depth];

        for (int[] dir : combs[curCctv.type]) {
            for (int d : dir) {
                hiddenArea -= change(curCctv.x, curCctv.y, d, 0, -(depth + 1));
            }

            recur(depth + 1, hiddenArea);

            for (int d : dir) {
                hiddenArea += change(curCctv.x, curCctv.y, d, -(depth + 1), 0);
            }
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine(), " ");
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        board = new int[n][m];
        int hiddenArea = 0;

        for (int x = 0; x < n; x++) {
            st = new StringTokenizer(br.readLine(), " ");

            for (int y = 0; y < m; y++) {
                board[x][y] = st.nextToken().charAt(0) - '0';

                if (board[x][y] == 0) {
                    hiddenArea++;
                } else if (board[x][y] <= 5) {
                    cctvs[cctvCount++] = new CCTV(x, y, board[x][y]);
                }
            }
        }

        minHiddenArea = hiddenArea;
        recur(0, hiddenArea);
        System.out.println(minHiddenArea);
    }

}

File: baekjoon_all/15000+/boj_15683.cpp
// Solve 2025-10-14

#include <bits/stdc++.h>

#define FASTIO ios_base::sync_with_stdio(false);cin.tie(NULL);
#define ALL(v) v.begin(),v.end()
#define UNIQUE(v) v.erase(unique(v.begin(),v.end()),v.end());
#define SETW(n, c) cout << setw(n) << setfill(c);
#define SETP(n) cout << fixed << setprecision(n);

using namespace std;
using ll = long long;
using uint = unsigned int;
using ull = unsigned long long;
using ld = long double;
using pii = pair<int, int>;
using pll = pair<ll, ll>;
const int INF = 0x3f3f3f3f;
const int MOD = 1000000007;

struct CCTV{
    int x, y, type;

    CCTV(int x, int y, int type) : x(x), y(y), type(type) {
    }
};

int n, m;
int board[8][8];
vector<CCTV> cctv;
int dx[4] = {-1, 0, 1, 0};
int dy[4] = {0, 1, 0, -1};
vector<vector<vector<int>>> cctv_direction = {
    {},
    {{0}, {1}, {2}, {3}},
    {{0, 2}, {1, 3}},
    {{0, 1}, {1, 2}, {2, 3}, {3, 0}},
    {{0, 1, 2}, {1, 2, 3}, {2, 3, 0}, {3, 0, 1}},
    {{0, 1, 2, 3}}
};
int mn;

int change(int x, int y, int d, int val, int new_val) {
    int ret = 0;
    x += dx[d];
    y += dy[d];

    while (x >= 0 && x < n && y >= 0 && y < m && board[x][y] != 6) {
        if (board[x][y] == val) {
            board[x][y] = new_val;
            ret++;
        }

        x += dx[d];
        y += dy[d];
    }

    return ret;
}

void recur(int depth, int val) {
    if (depth == size(cctv)) {
        mn = min(mn, val);
        return;
    }

    CCTV& cur_cctv = cctv[depth];

    for (vector<int>& dir : cctv_direction[cur_cctv.type]) {
        for (int d : dir) {
            val -= change(cur_cctv.x, cur_cctv.y, d, 0, -(depth + 1));
        }

        recur(depth + 1, val);

        for (int d : dir) {
            val += change(cur_cctv.x, cur_cctv.y, d, -(depth + 1), 0);
        }
    }
}

int main() {
    FASTIO;

    cin >> n >> m;
    int val = 0;

    for (int x = 0; x < n; x++) {
        for (int y = 0; y < m; y++) {
            cin >> board[x][y];

            if (board[x][y] == 0) {
                val++;
            }
            else if (board[x][y] != 6) {
                cctv.emplace_back(x, y, board[x][y]);
            }
        }
    }

    mn = val;
    recur(0, val);
    cout << mn << '\n';
    return 0;
}
